partner_preview_scripts/ValidationLab.py

```diff
@@ -3,432 +3,472 @@
+            """This function returns a struct with the biomarker name and description."""
+            # If the biomarker is not applied on the given cell type the value is zero:
+            if biomarker == "0":
+                return None
+
+            # If the biomarker value has a direct mapping:
+            if "direct_mapping" in biomarker_map[column_name]:
+                try:
+                    return biomarker_map[column_name]["direct_mapping"][biomarker]
+                except KeyError:
+                    logging.warning(
+                        f"Could not find direct mapping for {column_name}:{biomarker}"
+                    )
+                    return None
+
+            # If the value needs to be parsed:
+            if biomarker == "wt":
+                return {
+                    "name": biomarker_map[column_name]["name"] + biomarker,
+                    "description": biomarker_map[column_name]["description"]
+                    + "wild type",
+                }
+            elif biomarker == "mut":
+                return {
+                    "name": biomarker_map[column_name]["name"] + biomarker,
+                    "description": biomarker_map[column_name]["description"] + "mutant",
+                }
+            else:
+                logging.warning(
+                    f"Could not find direct mapping for {column_name}: {biomarker}"
+                )
+                return None
+
+        return wrapped_function
+
+    @classmethod
+    def get_biomarkers(
+        cls: type[BiomarkerParser], biomarker_df: DataFrame
+    ) -> DataFrame:
+        biomarkers_in_data = [
+            biomarker
+            for biomarker in cls.BIOMARKERMAPS.keys()
+            if biomarker in biomarker_df.columns
+        ]
 
-        # The first column is the gene name, the rest are the hypotheses:
-        hypothesis_columns = hypothesis_df.columns[1:]
+        # Applying the full map on the dataframe one-by-one:
+        processed_biomarkers = reduce(
+            lambda DF, value: DF.withColumn(*value),
+            map(
+                # Function to process biomarker:
+                lambda biomarker: (
+                    biomarker,
+                    cls._get_biomarker_wrapper(cls.BIOMARKERMAPS)(
+                        f.lit(biomarker),
+                        f.col(biomarker),
+                    ),
+                ),
+                # Iterator to apply the function over:
+                biomarkers_in_data,
+            ),
+            biomarker_df,
+        )
 
-        unpivot_expression = f"""stack({len(hypothesis_columns)}, {", ".join([f"'{x}', `{x}`" for x in hypothesis_columns])} ) as (hypothesis, {call})"""
+        # The biomarker columns are unstacked into one single 'biomarkers' column:
+        biomarker_unstack = f"""stack({len(biomarkers_in_data)}, {", ".join([f"'{x}', {x}" for x in biomarkers_in_data])}) as (biomarker_name, biomarker)"""
```

Comment:
Is a custom unstacking fucntion necessary or could we use `unpivot`? https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.DataFrame.unpivot.html

Reply:
When I learnt how to do unpivot, this functionality was not around. Also the linked docs says: `New in version 3.4.0.` so not sure if we can rely on this.